pkg/deviceshifu/deviceshifuhttp/deviceshifuhttp_test.go

```diff
@@ -186,3 +195,236 @@ func TestCreatehttpURIStringNoQuery(t *testing.T) {
+			instruction: "test_instruction",
+			properties:  mockDeviceShifuInstruction(),
+		},
+	}
+
+	ds := mockDeviceServer(mockHandlerHTTPCli, t)
+	defer ds.Close()
+	dc := mockRestClient(ds.URL, "testing")
+
+	// start device client testing
+	r := dc.Post().Param("timeout", "1").Param("stub_toleration", "1").Do(context.TODO())
+	assert.Nil(t, r.Error())
+
+	r = dc.Post().Param("timeout", "aa").Param("stub_toleration", "1").Do(context.TODO())
+	assert.Equal(t, "the server rejected our request for an unknown reason", r.Error().Error())
+
+	r = dc.Post().Param("timeout", "-1").Param("stub_toleration", "aa").Do(context.TODO())
+	assert.Equal(t, "the server rejected our request for an unknown reason", r.Error().Error())
+
+}
+
+func mockRestClient(host string, path string) *rest.RESTClient {
+	c, err := rest.RESTClientFor(
+		&rest.Config{
+			Host:    host,
+			APIPath: path,
+			ContentConfig: rest.ContentConfig{
+				GroupVersion:         &v1.SchemeGroupVersion,
+				NegotiatedSerializer: scheme.Codecs.WithoutConversion(),
+			},
+		},
+	)
+	if err != nil {
+		klog.Errorf("mock client for host %s, apipath: %s failed,", host, path)
+		return nil
+	}
+
+	return c
+}
+
+type MockCommandHandler interface {
+	commandHandleFunc() http.HandlerFunc
+}
+
+func mockDeviceServer(h MockCommandHandler, t *testing.T) *httptest.Server {
+	// catch device http request and response properly with specific paths
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		path := r.URL.Path
+		switch path {
+		case "/testing/apps/v1":
+			klog.Info("ds get testing call, calling the handler server")
+			assert.Equal(t, "/testing/apps/v1", path)
+			f := h.commandHandleFunc()
+			f.ServeHTTP(w, r)
+		default:
+			w.Header().Add("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			klog.Info("ds default request, path:", path)
+		}
+	}))
+	return server
+}
+
+func mockHandlerServer(t *testing.T) *httptest.Server {
+	// catch handler http request and response properly with specific paths
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		path := r.URL.Path
+		switch path {
+		case "/test_instruction":
+			w.Header().Add("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			assert.Equal(t, "/test_instruction", path)
+			klog.Info("handler get the instruction and executed.")
+		default:
+			w.WriteHeader(http.StatusOK)
+			klog.Info("hs get default request, path:", path)
+		}
+
+	}))
+	return server
+}
+
+func mockDeviceShifuInstruction() *deviceshifubase.DeviceShifuInstruction {
+	return &deviceshifubase.DeviceShifuInstruction{
+		DeviceShifuInstructionProperties: []deviceshifubase.DeviceShifuInstructionProperty{
+			{
+				ValueType:    "testing",
+				ReadWrite:    "rw",
+				DefaultValue: "0",
+			},
+		},
+		DeviceShifuProtocolProperties: map[string]string{
+			"test_key": "test_value",
+		},
+	}
+}
+
+func Test_collectHTTPTelemtries(t *testing.T) {
```

Comment:
should we follow the same camelcase naming scheme ?

Reply:
This was my old convention for testing a package private method in some old version IDE as I remember.
I checked in current VS code, now it should be deprecated. will change.